### PR TITLE
refactor!: Use different Id types

### PR DESCRIPTION
**Pull request by aawsome**
Breaking change:
Use different `Id` types for all places where `Id`s play a role. This makes the source code much more understandable.

This also prevents some nasty kind of bugs. This was possible before:
```rust
let snap = repo.get_snapshot_from_str("latest", |_| true)?;
let node = repo.node_from_path(snap.id, "/my/path".into())?;
```
But this wasn't working as `snap.id` is a snapshot id and `node_from_path` requires a tree id. This code will now fail at compile time.

### CODE REVIEW

**Comment by aawsome** [comment]
@simonsan I think this is ready for a review. I did not finish changing all internals where `PackedId` can be used, but the interface should now just use the Id newtypes.

File: crates/core/src/backend/decrypt.rs [hunk 1]
```diff
@@ -31,6 +31,8 @@ pub trait DecryptFullBackend: DecryptWriteBackend + DecryptReadBackend {}
 
 impl<T: DecryptWriteBackend + DecryptReadBackend> DecryptFullBackend for T {}
 
+pub type StreamResult<Id, F> = RusticResult<Receiver<RusticResult<(Id, F)>>>;
```
Comment (simonsan) [inline crates/core/src/backend/decrypt.rs:34]: Do we need to export this for public API?

File: crates/core/src/blob.rs [hunk 2]
```diff
@@ -66,17 +66,46 @@ impl<T: Default> Initialize<T> for BlobTypeMap<T> {
     }
 }
 
+new_id!(BlobId, "blob");
+
+pub trait PackedId: Copy + Into<BlobId> + From<BlobId> {
```
Comment (simonsan) [inline crates/core/src/blob.rs:71]: Also here, does it need to be part of the public API?

File: crates/core/src/blob.rs [hunk 3]
```diff
@@ -66,17 +66,46 @@ impl<T: Default> Initialize<T> for BlobTypeMap<T> {
...
 /// A `Blob` is a file that is stored in the backend.
 ///
 /// It can be a `tree` or a `data` blob.
 ///
 /// A `tree` blob is a file that contains a list of other blobs.
 /// A `data` blob is a file that contains the actual data.
-#[derive(Debug, PartialEq, Eq, Clone, Constructor)]
-pub(crate) struct Blob {
+#[derive(Debug, PartialEq, Eq, Copy, Clone, Constructor)]
+pub struct Blob {
```
Comment (simonsan) [inline crates/core/src/blob.rs:105]: Part of public API?

File: crates/core/src/blob/packer.rs [hunk 4]
```diff
@@ -20,16 +20,17 @@ use crate::{
...
         configfile::ConfigFile,
         indexfile::{IndexBlob, IndexPack},
-        packfile::{PackHeaderLength, PackHeaderRef},
+        packfile::{PackHeaderLength, PackHeaderRef, PackId},
         snapshotfile::SnapshotSummary,
     },
 };
 
+use super::BlobId;
+
```
Comment (simonsan) [inline crates/core/src/blob/packer.rs:33]: Relative import

File: crates/core/src/blob.rs [hunk 5]
```diff
@@ -66,17 +66,46 @@ impl<T: Default> Initialize<T> for BlobTypeMap<T> {
...
 
+new_id!(BlobId, "blob");
+
+pub trait PackedId: Copy + Into<BlobId> + From<BlobId> {
+    const TYPE: BlobType;
+}
+
+#[macro_export]
+/// Generate newtypes for `Id`s identifying packed blobs
+macro_rules! new_blobid {
```
Comment (simonsan) [inline crates/core/src/blob.rs:77]: Would name it something like `impl_blob_id` to make it easier readable and make people aware, that this just adds `impl` to a type.

File: crates/core/src/id.rs [hunk 6]
```diff
@@ -16,6 +16,33 @@ pub(super) mod constants {
     pub(super) const HEX_LEN: usize = LEN * 2;
 }
 
+#[macro_export]
+/// Generate newtypes for `Id`s identifying Repository files
+macro_rules! new_id {
```
Comment (simonsan) [inline crates/core/src/id.rs:21]: Would name it `define_new_id_struct`, so it's clear, here is a new struct being defined.

File: crates/core/src/repofile.rs [hunk 7]
```diff
@@ -10,9 +12,41 @@ pub(crate) mod snapshotfile;
...
+
+/// Marker trait for Ids which identify repository files
+pub trait RepoId: Deref<Target = Id> + From<Id> + Sized + Send + Sync + 'static {
+    /// The [`FileType`] associated with Id type
+    const TYPE: FileType;
+}
+
+#[macro_export]
+/// Generate newtypes for `Id`s identifying Repository files
+macro_rules! new_repoid {
```
Comment (simonsan) [inline crates/core/src/repofile.rs:27]: Would name it `impl_repo_id` to make us aware, that it just add an implementation for RepoId

File: crates/core/src/repofile.rs [hunk 8]
```diff
@@ -10,9 +12,41 @@ pub(crate) mod snapshotfile;
...
+    ($a:ident, $b: expr) => {
+        $crate::new_id!($a, concat!("repository file of type", stringify!($b)));
+        impl $crate::repofile::RepoId for $a {
+            const TYPE: FileType = $b;
+        }
+    };
+}
+
+#[macro_export]
```
Comment (simonsan) [inline crates/core/src/repofile.rs:36]: Would name it `impl_repo_file` to make us aware, that it just adds an implementation for RepoFile

**Review by simonsan — CHANGES_REQUESTED** [review]
Great job! That looks so much cleaner now! A few nits:

File: crates/core/src/blob.rs (hunk 2 above)
Comment (aawsome) [reply crates/core/src/blob.rs:71]: Yes, it is used in `Repository::get_index_item`.

File: crates/core/src/backend/decrypt.rs (hunk 1 above)
Comment (aawsome) [reply crates/core/src/backend/decrypt.rs]: no, is changed.

File: crates/core/src/blob.rs (hunk 3 above)
Comment (aawsome) [reply crates/core/src/blob.rs]: no, is changed.

File: crates/core/src/blob/packer.rs (hunk 4 above)
Comment (aawsome) [reply crates/core/src/blob/packer.rs]: is fixed.

File: crates/core/src/blob.rs (hunk 5 above)
Comment (aawsome) [reply crates/core/src/blob.rs]: renamed.

File: crates/core/src/id.rs (hunk 6 above)
Comment (aawsome) [reply crates/core/src/id.rs]: renamed.

File: crates/core/src/repofile.rs (hunk 7 above)
Comment (aawsome) [reply crates/core/src/repofile.rs]: renamed.

File: crates/core/src/repofile.rs (hunk 8 above)
Comment (aawsome) [reply crates/core/src/repofile.rs:36]: renamed.

File: crates/core/src/blob.rs (hunk 2 above)
Comment (simonsan) [reply crates/core/src/blob.rs:71]: Do we then need to export it directly in `lib.rs`?

**Review by simonsan — APPROVED** [review]

**Pull request merged**